♻️ REFACTOR: Reduce JS/CSS size and improve accessibility

foster999 opened the pull request:
Removes:
* semantic-ui assets
* unused 'sphinx_tabs_nowarn' sphinx option

Changes:
* tab structure to use [tab roles](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Roles/Tab_Role)
* tab label colour, to increase contrast with background
* assets are now copied across by `sphinx` rather than explicitly in the extension

Adds:
* `tabindex` atributes to allow users to focus and switch tabs using a keyboard (left and right arrows, and enter key to change)
* ARIA labels for tabs and panels
* ability to close tabs panel by clicking open tab
* persistence of last selected group or code tab during the session (when session storage is permitted)
* margin below images inside tab content

Closes #92, closes #66, closes #87, closes #90 and closes #80, closes #65, closes #93.
It also includes an equivalent of #73

foster999 commented:
Tablist bottom border can appear half-covered. Try only setting tab background colour when selected

foster999 commented:
Quite a large change to the underlying HTML here, so wouldn't be surprised if it introduces bugs for some use cases... Though I've trialed it with the demos in the docs and it seems to work well.

Also worth noting that:
1. Tabs and tab panels will now be focused on clicking, giving them a black outline until another element is focused. This focus highlighting is useful for those navigating by keyboard, so shouldn't be hidden.
2. Navigating across a tablist using arrow keys doesn't currently work if there are next/previous buttons on the page (i.e. `jupyter-book`s), because of an event listener in `sphinx`'s `doctools.js` triggers the page to change. I've raised a separate PR to disable this when `button` elements are focused, like those used for tabs here.

pradyunsg commented:
Ahoy! As a heads up, I won't be properly reviewing this myself anytime soon (context [here](https://github.com/executablebooks/meta/issues/145#issuecomment-708667465)) but I definitely wanted to drop in and say -- Thanks a lot for picking this up, and as far as I can see, this looks great!

foster999 commented:
Hey @chrisjsewell, please could you provide a review of this sometime? I appreciate that there's quite a few changes piled in here, but it made sense to include them given the extent of the rewrite.

If there's nothing obviously broken here I'll be around to pick up any bugs I may have thrown in, if released 😅

Daltz333 commented:
There seems to be some weird white outline of the tabbed content that just doesn't fit. 
![image](https://user-images.githubusercontent.com/10674555/96348792-5189ba80-1079-11eb-8191-9fe0f86e7ba1.png)

foster999 commented:
@Daltz333 thanks for spotting that - I'd only tested with a white background... Only the open tab should have a background now

chrisjsewell commented:
Heya soz for the delayed response. I will try to have a look at this in the next few days

foster999 commented:
No rush at all 👍

Daltz333 commented:
@chrisjsewell Any update on this? I'd love to use sphinx-tabs with sphinx-panels.

chrisjsewell commented on sphinx_tabs/static/doctools.js:
```diff
@@ -0,0 +1,316 @@
+/*
```
I'm a little tentative on overriding core aspects of sphinx 😬 (in case it is changed in some way upstream) are the modifications actually specific to sphinx-tabs or could you make a PR to sphinx to add these modifications?
At the least could you add a link to where the original document can be viewed (on https://github.com/sphinx-doc/sphinx) and clearly comment where the modifications are and why they have been made.

Daltz333 replied:
Some pretty short research shows that it's near identical to https://github.com/sphinx-doc/sphinx/blob/3.x/sphinx/themes/basic/static/doctools.js

Daltz333 replied:
Actually, it looks like they are identical, not sure why it's overriding?

foster999 replied:
I'd been using this fork, so it was meant as a short term workaround really. But sphinx-doc/sphinx#8316 included the minor change in sphinx, so I'll drop it here 👍

foster999 replied:
Removed this file from the PR

chrisjsewell commented:
Heya, @Daltz333 I don't want to be too much of a blocker on this package, so I've invited you and @foster999 as maintainers

I've made a quick review comment above also:

> It also includes an equivalent of #73

Could you add an example of this to the documentation (https://sphinx-tabs--94.org.readthedocs.build/en/94/#), and in fact it would be good to improve this page with a bit more explanation of the extension, including sphinx configuration options etc

foster999 commented:
Thanks @chrisjsewell, sounds good to me

I've dropped the sphinx js override and updated the RTD docs to include code examples and describe features. Please could one of you have a quick look at https://sphinx-tabs--94.org.readthedocs.build/en/94/#, then I think it should be good to go 😄

Daltz333 reviewed (approved):
Looks fine to me! I think we should ship it if no other objections!

chrisjsewell commented on docs/index.rst:
````diff
@@ -1,6 +1,59 @@
+Installation
+============
+
+```bash
+pip install sphinx-tabs
+```
+
+To enable the extension in Sphinx, add the following to your conf.py:
+
+```python
+extensions = ['sphinx_tabs.tabs']
+```
+
+If needed, there is a configuration option to allow additional builders to be considered compatible. For example, to add the `linkcheck` builder, add the following to your conf.py:
+
+```python
+sphinx_tabs_valid_builders = ['linkcheck']
+```
+
````
you are trying to use markdown in an RST file here 😉 

should be e.g.

```
.. code-block:: bash

   pip install sphinx-tabs

```

Daltz333 replied:
Ah, I didn't notice, as it renders partially correct on the actual site.
![image](https://user-images.githubusercontent.com/10674555/105637698-b43edf00-5e3c-11eb-81e5-70fd559ea4b9.png)

chrisjsewell replied:
yeh I know, I just notice the extra `` ` `` and though something was a miss

foster999 replied:
🤦🏻‍♂️ I copied that straight from the README without thinking...

chrisjsewell reviewed (changes requested):
thanks @foster999 looking good, just a minor thing to change

chrisjsewell commented:
cheers!

chrisjsewell commented:
I imagine this should be released as v2.0.0

pradyunsg commented:
Hurray! Thanks for doing this folks! ^>^

foster999 commented:
Cheers for the review both! Shall punch a new release now

chrisjsewell commented:
I assume you know how: https://github.com/executablebooks/.github/blob/master/CONTRIBUTING.md#the-process-of-creating-a-release
